### PR TITLE
MODFISTO-41 - Data migration scripts for recent schema changes

### PR DESCRIPTION
**Pull request by khandramai**
[MODFISTO-41](https://issues.folio.org/browse/MODFISTO-41)

Instances:
- before: [original.txt](https://github.com/folio-org/mod-finance-storage/files/3744216/original.txt)
- after: [updated.txt](https://github.com/folio-org/mod-finance-storage/files/3744217/updated.txt)

## Purpose
data migration scripts for all breaking schema changes starting after the Q3.2 release
## 
#### TODOS and Open Questions
- Default values for currency stetted to USD at the moment.

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/resources/templates/db_scripts/transactions.sql [hunk 1]
```diff
@@ -117,3 +117,74 @@ $recalculate_totals$ LANGUAGE plpgsql;
...
+	WHERE jsonb::json ->> 'awaitingPayment' iS NOT NULL;
+
+UPDATE ${myuniversity}_${mymodule}.transaction
+SET jsonb =
+	jsonb #- '{encumbered}' || jsonb_build_object('initialAmountEncumbered', jsonb::json -> 'encumbered')
+	WHERE jsonb::json ->> 'encumbered' iS NOT NULL;
+
+UPDATE ${myuniversity}_${mymodule}.transaction
+SET jsonb =
+	jsonb #- '{expenditures}' || jsonb_build_object('amountExpended', jsonb::json -> 'expenditures')
```
Comment (pautau-s) [inline src/main/resources/templates/db_scripts/transactions.sql]: `amountExpended` should be a part of `encumbrance` sub-object

File: src/main/resources/templates/db_scripts/transactions.sql [hunk 2]
```diff
@@ -117,3 +117,74 @@ $recalculate_totals$ LANGUAGE plpgsql;
 
 DROP TRIGGER IF EXISTS recalculate_totals on ${myuniversity}_${mymodule}."transaction";
 CREATE TRIGGER recalculate_totals AFTER INSERT ON "transaction" FOR EACH ROW EXECUTE PROCEDURE ${myuniversity}_${mymodule}.recalculate_totals();
+
+-- Rename fields
+
+UPDATE ${myuniversity}_${mymodule}.transaction
+SET jsonb =
+	jsonb #- '{awaitingPayment}' || jsonb_build_object('amountAwaitingPayment', jsonb::json -> 'awaitingPayment')
```
Comment (pautau-s) [inline src/main/resources/templates/db_scripts/transactions.sql]: `amountAwaitingPayment` should be a part of `encumbrance` sub-object

**Comment by javalkar-varun** [comment]
How did you test that these scripts are working?

File: src/main/resources/templates/db_scripts/migration/budgets.sql [hunk 3]
```diff
@@ -0,0 +1,50 @@
+-- Rename fields
+
+UPDATE ${myuniversity}_${mymodule}.budget
+SET jsonb =
+	jsonb #- '{limitEncPercent}' || jsonb_build_object('allowableEncumbrance', jsonb::json ->> 'limitEncPercent')
+	WHERE jsonb::json->>'limitEncPercent' iS NOT NULL;
+
+UPDATE ${myuniversity}_${mymodule}.budget
+SET jsonb =
+	jsonb #- '{limitExpPercenet}' || jsonb_build_object('allowableExpenditure', jsonb::json ->> 'limitExpPercenet')
```
Comment (javalkar-varun) [inline src/main/resources/templates/db_scripts/migration/budgets.sql]: Typo `'{limitExpPercenet}'` -> `'{limitExpPercent}'`

**Comment by khandramai** [comment]
> How did you test that these scripts are working?

1) Checkout and build appropriate tag
2) Deploy old version of module
3) Create new test tenant
4) Enable old module version for test tenant
5) Populate sample data for old module
6) Build actual module version
7) Update old module version to new one

Additional information can be found here https://github.com/folio-org/mod-orders-storage/pull/157

**Comment by khandramai** [comment]
After last changes: 
[original_schemas.txt](https://github.com/folio-org/mod-finance-storage/files/3778820/original_schemas.txt)
[updated_schemas.txt](https://github.com/folio-org/mod-finance-storage/files/3778821/updated_schemas.txt)

File: src/main/resources/templates/db_scripts/migration/budgets.sql [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  AND jsonb::json ->> 'budgetStatus' <> 'Closed');
+
+-- Change type of non-required fields
+
+UPDATE ${myuniversity}_${mymodule}.budget
+SET jsonb =
+   jsonb || jsonb_build_object('tags', jsonb_build_object('tagList', jsonb::json -> 'tags'))
+   WHERE (jsonb::json #> '{tags, tagList}') iS NULL;
+
+-- Make fields required
```
[the remark this answers is not in the document]
Comment (craigmcnally) [reply src/main/resources/templates/db_scripts/migration/budgets.sql]: a budget is by definition a fund/fy, so not having that information somewhere in the budget is highly unlikely.  

This isn't ideal, but if the fundId and/or fiscalYearId fields aren't present, we can  look at "code" if it exists and try to parse that and determine the fund and fiscal year... 

e.g. if budget.code = HIST-FY19, look for fund.code = HIST and fiscalYear.cod = FY19.

Aside from that there really isn't much we can do about it... we can't guess the fund or fiscal year.  I'll reach out to Dennis to see what he has to say

[the remark this answers is not in the document]
Comment (khandramai) [reply src/main/resources/templates/db_scripts/migration/budgets.sql]: As per discussion with Team - budget without fiscal year Id and fund Id is highly unlikely so let it as is without additional actions.

**Pull request merged**